devito/ir/support/utils.py

```diff
@@ -209,12 +209,15 @@ def group_expressions(exprs):
     # Partion based on data dependences
     mapper = OrderedDict()
     ngroups = 0
-    for i, e1 in enumerate(exprs):
+    for e1 in exprs:
         if e1 in mapper:
+            # Optimization: we know already that a group for `e1` has been found
             continue
         found = False
-        for e2 in exprs[i+1:]:
-            if Scope([e1, e2]).has_dep:
+        for e2 in exprs:
+            if e1 is e2:
+                continue
+            elif Scope([e1, e2]).has_dep:
```

Comment:
Seems like a symmetric relationship to me. If yes, isn't the old way of starting from `i+1` better?

Reply:
that's what I thought originally (transitive in particular), but no because I'm updating ``mapper`` ASAP, so you may have
A -> B so both A and B are in mapper
then if you also have C and A -> C, you need to check A again